packages/client/lib/sync/txpool.ts

```diff
@@ -165,16 +177,38 @@ export class TxPool {
    * Adds a tx to the pool.
    *
    * If there is a tx in the pool with the same address and
-   * nonce it will be replaced by the new tx.
+   * nonce it will be replaced by the new tx, if it has a sufficent gas bump.
+   * (TODO) verifies certain constraints, if these are not matches, throws.
    * @param tx Transaction
    */
-  add(tx: TypedTransaction) {
+  async add(tx: TypedTransaction) {
     const sender: UnprefixedAddress = tx.getSenderAddress().toString().slice(2)
     const inPool = this.pool.get(sender)
     let add: TxPoolObject[] = []
     if (inPool) {
       // Replace pooled txs with the same nonce
-      add = inPool.filter((poolObj) => !poolObj.tx.nonce.eq(tx.nonce))
+      let existingTxn = inPool.find((poolObj) => poolObj.tx.nonce.eq(tx.nonce))
+      if (existingTxn) {
+        let currentGasPrice
+        switch (existingTxn.tx.type) {
+          case 0: 
+            currentGasPrice = (existingTxn.tx as Transaction).gasPrice
+            break
+          case 1: 
+            currentGasPrice = (existingTxn.tx as AccessListEIP2930Transaction).gasPrice
+            break
+          case 2:
+            currentGasPrice = (existingTxn.tx as FeeMarketEIP1559Transaction).maxFeePerGas
+            break
+          default: 
+            throw new Error("unknown tx type") // TODO make this error better
+        }
+        let minimumGasPrice = currentGasPrice.add(currentGasPrice.muln(this.minGasPriceBump).divn(100))
```

Comment:
hm that's a lot of boilerplate code just to get the tx gasPrice. would be nice to have a tx helper for this in the base tx class, can use the tx.supports(Capability) API to return the right response? (basically maxFeePerGas if FeeMarket, else gasPrice)

Reply:
I had exactly the same idea, I just took a look how we do this in VM, we could refactor [there](https://github.com/ethereumjs/ethereumjs-monorepo/blob/07a6d2e8a672d6bebad57a1e9df80c3e735fbf2e/packages/vm/src/runTx.ts#L356) (this gasPrice propagates via `TxContext` to `EVM` to `EEI`, where it is used to get the gas price in the `GASPRICE` opcode) as well, the only thing is that the helper needs a `baseFee` parameter in case of a 1559 txn.

Something like `getGasPrice(baseFee?: BN)` on `BaseTransaction`, on 1559 we know that we are in a 1559 block which thus has `baseFee` and otherwise in 2930/Legacy txs we can just make it `undefined` since it points to `gasPrice`. (This whole `baseFee` stuff also makes TxPool more interesting since you could include a tx in the pool which could be mined now, but at some later blocks not if `baseFee` rises above the `maxPriorityFeePerGas` ... :thinking: )

Reply:
Yeah that would be great!

The baseFee does add an interesting dynamic, so I don't think we should reject txs if they are below the baseFee, since it could eventually lower, but it would be nice on the rpc layer to know if a tx I am submitting is severely underpriced (meaning, it has very little chance to be mined). To start we can probably just model around how geth is already thinking about it, rather than coming up with our own rational, to best match their tx pool credentials.